src/main/java/com/hubspot/ringleader/watcher/WatcherFactory.java

```diff
@@ -29,26 +30,31 @@
   private final Object watcherLock;
   private final AtomicReference<CuratorFramework> curatorReference;
   private final AtomicLong curatorTimestamp;
+  private final AtomicBoolean replacementScheduled;
   private final AtomicBoolean started;
   private final AtomicBoolean closed;
-  private final AtomicInteger watchers;
+  private final Map<PersistentWatcher, Boolean> watchers;
```

Comment:
Keeping track of the watchers allows WatcherFactory to know which ones need to be refetched when the connection is replaced. This is easier than having the watchers themselves pass a Runnable - particularly as there could be many watchers sharing the same factory/curator.